core/state_update.go

```diff
@@ -51,38 +52,91 @@ func (d *StateDiff) Length() uint64 {
 }
 
 func (d *StateDiff) Hash() *felt.Felt {
-	prefixHash, err := new(felt.Felt).SetString("STARKNET_STATE_DIFF0")
-	if err != nil {
-		// shouldn't happen
-		panic(err)
-	}
+	var digest crypto.PoseidonDigest
+
+	digest.Update(new(felt.Felt).SetBytes([]byte("STARKNET_STATE_DIFF0")))
 
-	hashElems := []*felt.Felt{prefixHash}
+	// updated_contracts = deployedContracts + replacedClasses
+	// Digest: [number_of_updated_contracts, address_0, class_hash_0, address_1, class_hash_1, ...].
+	numOfUpdatedContracts := uint64(len(d.DeployedContracts) + len(d.ReplacedClasses))
+	digest.Update(new(felt.Felt).SetUint64(numOfUpdatedContracts))
 
-	// updated contracts
-	numOfUpdatedContracts := len(d.DeployedContracts) + len(d.ReplacedClasses)
-	hashElems = append(hashElems, new(felt.Felt).SetUint64(uint64(numOfUpdatedContracts)))
-	// todo sort
-	for addr, classHash := range d.DeployedContracts {
-		hashElems = append(hashElems, &addr, classHash)
+	// The sequencer guarantees that a contract cannot be:
+	// - deployed twice,
+	// - deployed and have its class replaced in the same state diff, or
+	// - have its class replaced multiple times in the same state diff.
+	updatedContracts := make(map[felt.Felt]*felt.Felt)
+	for k, v := range d.DeployedContracts {
+		updatedContracts[k] = v
 	}
-	// todo sort
-	for addr, classHash := range d.ReplacedClasses {
-		hashElems = append(hashElems, &addr, classHash)
+	for k, v := range d.ReplacedClasses {
+		updatedContracts[k] = v
```

Comment:
maybe `maps.Copy(updatedContracts, d.ReplacedClasses)` and the same for `d.DeployedContracts` above?

Reply:
Yeah will do